### PR TITLE
Tightening down php code sniffer

### PR DESCRIPTION
**Pull request by jaspernbrouwer**
This PR mainly tightens down the rules for PHP Code Sniffer, and addresses all issues that arise from it. Code style is now enforced a lot more strictly.

See the commit messages for a couple of small additional fixes relating to the same subject.

### CODE REVIEW

File: specs/Elewant/Herding/Application/Commands/FormHerdHandlerSpec.php [hunk 1]
```diff
@@ -12,9 +12,7 @@
 
 final class FormHerdHandlerSpec extends ObjectBehavior
 {
-    /**
-     * @var HerdCollection
-     */
+    /** @var HerdCollection */
```
Comment (ramondelafuente) [inline specs/Elewant/Herding/Application/Commands/FormHerdHandlerSpec.php]: Why this change? Is there a specific reason to require the short block?

File: src/Bundles/DevelopmentBundle/Security/DevelopmentOauthResourceOwner.php [hunk 2]
```diff
@@ -1,72 +1,64 @@
...
-        // @todo: Implement addPaths() method.
     }
 
-    /**
-     * @param string $refreshToken    Refresh token
-     * @param array  $extraParameters An array of parameters to add to the url
-     */
     public function refreshAccessToken($refreshToken, array $extraParameters = []): void
     {
-        // @todo: Implement refreshAccessToken() method.
```
Comment (ramondelafuente) [inline src/Bundles/DevelopmentBundle/Security/DevelopmentOauthResourceOwner.php:70]: If this is not a todo, what is it? Leaving an empty function might not be an improvement.

File: src/Bundles/UserBundle/Entity/User.php [hunk 3]
```diff
@@ -170,16 +169,18 @@ private function hasConnection(string $resource): bool
     /**
      * We don't use associations, so we can safely store the user in sessions.
      * The user-provider will refresh the user, to make it complete and managed.
+     *
+     * @return string
```
Comment (ramondelafuente) [inline src/Bundles/UserBundle/Entity/User.php:174]: Is this annotation required because we want an explanation in the docblock? Because the return value of the method is good enough for me.

File: src/Elewant/Herding/Application/Commands/AbandonHerd.php [hunk 4]
```diff
@@ -20,6 +20,7 @@ public static function forShepherd(string $herdId, string $shepherdId): self
     }
 
     /**
+     * @return HerdId
```
Comment (ramondelafuente) [inline src/Elewant/Herding/Application/Commands/AbandonHerd.php:23]: Same here.. If PHPStorm could just stop complaining about missing return values that are declared in the code, that would be great. Mkay.

File: src/Elewant/Herding/Application/ProcessManagers/NewUserSubscriber.php [hunk 5]
```diff
@@ -24,22 +24,22 @@ final class NewUserSubscriber implements EventSubscriberInterface
...
 
     public function __construct(HerdRepository $herdRepository, CommandBus $commandBus)
     {
         $this->herdRepository = $herdRepository;
-        $this->commandBus     = $commandBus;
+        $this->commandBus = $commandBus;
     }
 
+    /**
+     * @return mixed[]
```
Comment (ramondelafuente) [inline src/Elewant/Herding/Application/ProcessManagers/NewUserSubscriber.php:43]: Yuck. Can't we point at a parent docblock with `@inherited` instead? :D

File: src/Elewant/Herding/DomainModel/Breed/Breed.php [hunk 6]
```diff
@@ -6,106 +6,108 @@
...
 
+    /**
+     * @return string[]
+     */
     public static function availableTypes(): array
     {
-        /** @noinspection PhpUnhandledExceptionInspection */
-        $reflected  = new ReflectionClass(self::class);
-        $validTypes = $reflected->getConstants();
+        try {
```
Comment (ramondelafuente) [inline src/Elewant/Herding/DomainModel/Breed/Breed.php:99]: What exactly are we protecting ourselfves from here? This try-catch will never ever catch anything.

(by the way, I think the basic problem with this code is me writing crap - I should just have made static constructors and not bothered with the constants)

File: src/Elewant/Herding/DomainModel/Breed/BreedCollection.php [hunk 7]
```diff
@@ -91,7 +101,7 @@ public function equals(BreedCollection $otherCollection): bool
     {
         $diff = array_diff($this->breeds, $otherCollection->breeds);
 
-        return empty($diff) && count($this->breeds) === count($otherCollection->breeds);
+        return !$diff && count($this->breeds) === count($otherCollection->breeds);
```
Comment (ramondelafuente) [inline src/Elewant/Herding/DomainModel/Breed/BreedCollection.php:107]: Since array_diff returns an array, what is wront with `empty()`? Just curious..

File: src/Elewant/Herding/DomainModel/Herd/Herd.php [hunk 8]
```diff
@@ -278,10 +260,13 @@ private function applyAnElePHPantWasAdoptedByHerd(ElePHPantId $elePHPantId, Bree
     private function applyAnElePHPantWasAbandonedByHerd(ElePHPantId $elePHPantId, Breed $breed): void
     {
         $this->breeds->remove($breed);
+
         foreach ($this->elePHPants as $key => $elePHPant) {
-            if ($elePHPant->elePHPantId()->equals($elePHPantId)) {
-                unset($this->elePHPants[$key]);
+            if (!$elePHPant->elePHPantId()->equals($elePHPantId)) {
```
Comment (ramondelafuente) [inline src/Elewant/Herding/DomainModel/Herd/Herd.php:279]: I don't quite agree with the return early here, since we loop over all the elements anyways.

`foreach() { if (true) {do something }}`
is more comprehensible than
`foreach() { if (!true) { continue } do something}` 

But it's also not a big thing.

File: src/Elewant/Reporting/Infrastructure/Doctrine/Statistics/NumberOfRepository.php [hunk 9]
```diff
@@ -60,25 +61,26 @@ public function herdsEverFormed(): int
...
+     * @param DateTimeInterface $to
      * @return int
-     * @throws \Doctrine\ORM\NonUniqueResultException
+     * @throws NonUniqueResultException
      */
-    public function newElePHPantsAdoptedBetween(\DateTimeInterface $from, \DateTimeInterface $to): int
+    public function newElePHPantsAdoptedBetween(DateTimeInterface $from, DateTimeInterface $to): int
     {
-        $dql = <<<EOQ
+        $dql = <<<'EOQ'
```
Comment (ramondelafuente) [inline src/Elewant/Reporting/Infrastructure/Doctrine/Statistics/NumberOfRepository.php:73]: We can use indentation in heredocs now! 

```
$dql = <<<'EOQ'
           SELECT COUNT(e)
           FROM Herding:ElePHPant e
           WHERE e.adoptedOn BETWEEN :from AND :to
           EOQ;
```
Should we add that? (not in this PR, just a remark)

File: src/Elewant/Reporting/Infrastructure/SymfonyConsole/HerdingStatisticsCommand.php [hunk 10]
```diff
@@ -81,14 +76,13 @@ protected function execute(InputInterface $input, OutputInterface $output): void
...
      * @param string|null $inputTo
-     *
      * @return DateTimeImmutable[]
      * @throws Exception
      */
     private function prepareDateArguments(?string $inputFrom, ?string $inputTo): array
     {
         $from = new DateTimeImmutable($inputFrom ?? 'monday last week');
-        $to   = ($inputTo === null) ? $from->modify('+6 days') : new DateTimeImmutable($inputTo);
+        $to = $inputTo === null ? $from->modify('+6 days') : new DateTimeImmutable($inputTo);
```
Comment (ramondelafuente) [inline src/Elewant/Reporting/Infrastructure/SymfonyConsole/HerdingStatisticsCommand.php]: I prefer the brackets on the first part of the ternary to  make is more obvious. Probably a code-style violation, but I was never as strict as you when it came to that stuff :)

File: src/Elewant/Webapp/Infrastructure/Doctrine/BreedMappingType.php [hunk 11]
```diff
@@ -74,7 +77,9 @@ public function convertToDatabaseValue($value, AbstractPlatform $platform): ?str
     }
 
     /**
-     * @inheritdoc
+     * @param AbstractPlatform $platform
+     * @return bool
+     * @phpcsSuppress SlevomatCodingStandard.Functions.UnusedParameter.UnusedParameter
```
Comment (ramondelafuente) [inline src/Elewant/Webapp/Infrastructure/Doctrine/BreedMappingType.php:82]: I've seen this a few times now and I'm triggered every time. I've decided I'm not a big fan of polluting the codebase with annotations meant for test-tooling... is there any way we can put these definitions outside the code?

File: src/Elewant/Webapp/Infrastructure/ProophProjections/HerdProjection.php [hunk 12]
```diff
@@ -20,13 +20,13 @@
...
 {
-
     public function project(ReadModelProjector $projector): ReadModelProjector
     {
         $projector->fromStream('event_stream')
             ->when(
                 [
-                    HerdWasFormed::class                  =>
+                    HerdWasFormed::class =>
+                    /** @phpcsSuppress SlevomatCodingStandard.Functions.UnusedParameter.UnusedParameter */
```
Comment (ramondelafuente) [inline src/Elewant/Webapp/Infrastructure/ProophProjections/HerdProjection.php:29]: Same here.. this is not really nice to deal with when writing code.

File: src/Elewant/Webapp/Infrastructure/ProophProjections/HerdReadModel.php [hunk 13]
```diff
@@ -150,33 +145,25 @@ public function onElePHPantWasAbandonedByHerd(ElePHPantId $elePHPantId): void
...
                 [
-                    'herd_id'    => $herdId->toString(),
-                    'breed'      => $breed->toString(),
+                    'herd_id' => $herdId->toString(),
+                    'breed' => $breed->toString(),
                     'desired_on' => $desiredOn->format('Y-m-d H:i:s'),
                 ]
             );
-        } catch (UniqueConstraintViolationException $e) {
+        } catch (DBALException $e) {
```
Comment (ramondelafuente) [inline src/Elewant/Webapp/Infrastructure/ProophProjections/HerdReadModel.php]: Does this mean the same thing? I wanted to skip duplicate-key errors specifically during rebuild of the read model. But this skips on any error...

File: src/Elewant/Webapp/Infrastructure/Twig/ComponentsExtension.php [hunk 14]
```diff
@@ -13,6 +13,7 @@
 
 final class ComponentsExtension extends AbstractExtension
 {
+    /** @return TwigFunction[] */
```
Comment (ramondelafuente) [inline src/Elewant/Webapp/Infrastructure/Twig/ComponentsExtension.php]: What's with the short notation on methods also? I can sort-of accept it for properties but this is confusing.

File: tests/Elewant/Webapp/Application/Controllers/ApiCommandAdoptElePHPantTest.php [hunk 15]
```diff
@@ -26,12 +26,12 @@ public function setUp()
         $this->client = $this->adoptElePHPant($this->herdId, Breed::blackAmsterdamphpRegular());
     }
 
-    public function test_command_adopt_elephpant_returns_http_status_202()
+    public function test_command_adopt_elephpant_returns_http_status_202(): void
```
Comment (ramondelafuente) [inline tests/Elewant/Webapp/Application/Controllers/ApiCommandAdoptElePHPantTest.php:31]: I really don't feel like having to "properly" code all the test functions. Tests do not require the same stricness as production code in my opinion.

File: tests/Elewant/Webapp/Application/Controllers/ApiCommandBase.php [hunk 16]
```diff
@@ -29,174 +30,144 @@ abstract class ApiCommandBase extends WebTestCase
...
-     */
-    private function getHerdListing(ContainerInterface $container)
+    private function getHerdListing(ContainerInterface $container): HerdListing
     {
         return $container->get('Elewant\Webapp\Infrastructure\ProophProjections\HerdListing');
     }
 
     /**
-     * This method can run a projection once
+     * This method can run a projection once.
```
Comment (ramondelafuente) [inline tests/Elewant/Webapp/Application/Controllers/ApiCommandBase.php:164]: Such improve. Much punctuation. So better!

**Review by ramondelafuente — APPROVED** [review]
There's a ton of stuff here, but none if it is blocking - just food for thought. Nice work!

File: specs/Elewant/Herding/Application/Commands/FormHerdHandlerSpec.php (hunk 1 above)
Comment (jaspernbrouwer) [reply specs/Elewant/Herding/Application/Commands/FormHerdHandlerSpec.php]: It was in the rules I found, but I agree it's a silly rule, I'm going to remove it.

File: src/Bundles/DevelopmentBundle/Security/DevelopmentOauthResourceOwner.php (hunk 2 above)
Comment (jaspernbrouwer) [reply src/Bundles/DevelopmentBundle/Security/DevelopmentOauthResourceOwner.php:70]: In this case it was unimplemented on purpose, but I'll leave a comment explaining that.

File: src/Elewant/Herding/Application/Commands/AbandonHerd.php (hunk 4 above)
Comment (jaspernbrouwer) [reply src/Elewant/Herding/Application/Commands/AbandonHerd.php]: Indeed, PhpStorm's inspections require these annotations when a docblock is present. There is an option that allows you to omit `@param` and `@return` when type-hints are in place, but they behave inconsistently.

Personally I'm in favor of having them and consistent, over not having them with weird exceptions. Plus I really like being able to "Inspect code" and having a clean list of what I need to fix.

File: src/Bundles/UserBundle/Entity/User.php (hunk 3 above)
Comment (jaspernbrouwer) [reply src/Bundles/UserBundle/Entity/User.php:174]: Indeed, PhpStorm's inspections require these annotations when a docblock is present. There is an option that allows you to omit `@param` and `@return` when type-hints are in place, but they behave inconsistently.

Personally I'm in favor of having them and consistent, over not having them with weird exceptions. Plus I really like being able to "Inspect code" and having a clean list of what I need to fix.

File: src/Elewant/Herding/Application/ProcessManagers/NewUserSubscriber.php (hunk 5 above)
Comment (jaspernbrouwer) [reply src/Elewant/Herding/Application/ProcessManagers/NewUserSubscriber.php:43]: This is regarding the rule `SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingTraversableReturnTypeHintSpecification`, which states that you should be more specific that `array` (or `Traversable`, etc). I like this rule, because in most cases you _can_ be more specific (`string[]` instead of `array` for example).

In this case, we _cannot_ be more specific, because several array-structures are allowed.

`@return mixed[]` is kind of a workaround, `mixed` is deemed specific enough. The alternative is suppressing the rule, but I want to minimize that.

PS: I'm not sure if phpcs looks at the parent in these cases. In this specific case the parent is `Symfony\Component\EventDispatcher\EventSubscriberInterface`, isn't more specific that `@return array`, so there is no "correct" parent to fall back to.

File: src/Elewant/Herding/DomainModel/Breed/Breed.php (hunk 6 above)
Comment (jaspernbrouwer) [reply src/Elewant/Herding/DomainModel/Breed/Breed.php:99]: I must confess that I'm not sure why I changed this. But I think it's because I'd prefer the try/catch over suppressing an inspection.

File: src/Elewant/Herding/DomainModel/Breed/BreedCollection.php (hunk 7 above)
Comment (jaspernbrouwer) [reply src/Elewant/Herding/DomainModel/Breed/BreedCollection.php:107]: `empty($diff)` has the same behavior as `!isset($diff) || !$diff` (or `isset($diff) === false || $diff == false`). Since we _know_ `isset($diff)` will return `true`, because `$diff` is defined in the line before the `if`, that `!isset($diff)` part is irrelevant. What remains then, is `!$diff`.

I can follow that `$diff` is an array (or list), and we want to test "is the list empty?", then `if (empty($list))` makes sense. But IMHO we're not speaking English here, but we're speaking PHP. So, if we really want to make "is the diff empty?" explicit in PHP, it would look something like `count($diff) === 0`. I think it's important to speak English when naming things (variables, methods, etc), but the actual implementation remains PHP.

Personally `if ($var)` or `if (!$var)` (i.e. does it evaluate to true or false) is fine. I don't really need to go as far as `if ($var === true)` or `if ($var === false)` (or `count($diff) === 0`). But hey! It's a matter of taste / personal preference, so I'm open to alternatives :slightly_smiling_face:

File: src/Elewant/Reporting/Infrastructure/Doctrine/Statistics/NumberOfRepository.php (hunk 9 above)
Comment (jaspernbrouwer) [reply src/Elewant/Reporting/Infrastructure/Doctrine/Statistics/NumberOfRepository.php:73]: Personally I like them without indentation. I usually use them when the content gets larger (above 2 or 3 lines), so make use of the full width as well.
But this opinion isn't that strong, so I'm fine either way!

File: src/Elewant/Reporting/Infrastructure/SymfonyConsole/HerdingStatisticsCommand.php (hunk 10 above)
Comment (jaspernbrouwer) [reply src/Elewant/Reporting/Infrastructure/SymfonyConsole/HerdingStatisticsCommand.php]: Yes, it violates the rule `SlevomatCodingStandard.PHP.UselessParentheses.UselessParentheses`.

I do prefer using parentheses when it improves readability. So I'll remove this rule.

File: src/Elewant/Webapp/Infrastructure/Doctrine/BreedMappingType.php (hunk 11 above)
Comment (jaspernbrouwer) [reply src/Elewant/Webapp/Infrastructure/Doctrine/BreedMappingType.php:82]: Yes, these suppressions can be done in the `phpcs.xml` file itself. For example:

```
<rule ref="PSR1.Methods.CamelCapsMethodName">
    <exclude-pattern>/(specs|tests)/</exclude-pattern>
</rule>
```

On one hand, I too dislike `@phpcsSuppress` (and `@noinspection`) in the code, but I dislike having to maintain them elsewhere even more.
Mainly because these suppressions tend to be added, but never removed. In my experience, nobody actually bothers to look at them and check if they're still needed.
So I prefer having them in the code where they tend to be maintained, over having them in `phpcs.xml` where they are forgotten.

File: src/Elewant/Herding/DomainModel/Breed/BreedCollection.php (hunk 7 above)
Comment (ramondelafuente) [reply src/Elewant/Herding/DomainModel/Breed/BreedCollection.php:107]: For me, it's a matter of exactly 0 bugs can come from the previous implementation with `empty`. As with `!diff`. So *shrug* and move on.

File: src/Elewant/Reporting/Infrastructure/Doctrine/Statistics/NumberOfRepository.php (hunk 9 above)
Comment (ramondelafuente) [reply src/Elewant/Reporting/Infrastructure/Doctrine/Statistics/NumberOfRepository.php:73]: Makes sense. I don't like the visual of the loss of indentation (it looks like a broken flow to me) but there is probably near-zero business value there so this is fine as-is.

File: src/Elewant/Webapp/Infrastructure/ProophProjections/HerdReadModel.php (hunk 13 above)
Comment (jaspernbrouwer) [reply src/Elewant/Webapp/Infrastructure/ProophProjections/HerdReadModel.php]: Hmm... I've seem to mistakenly concluded that `UniqueConstraintViolationException` isn't thrown anymore, but (since Doctrine ORM 2.6) it's hidden in some wrapping technique. I'll fix this so the original behavior is maintained.

File: src/Elewant/Webapp/Infrastructure/Doctrine/BreedMappingType.php (hunk 11 above)
Comment (ramondelafuente) [reply src/Elewant/Webapp/Infrastructure/Doctrine/BreedMappingType.php:82]: Sure.. I think it's likely to rot either way, but maybe less so when in the code. All it does is constantly remind me we're "improving" code in a bureaucratic way, rubber-stamping.

File: tests/Elewant/Webapp/Application/Controllers/ApiCommandAdoptElePHPantTest.php (hunk 15 above)
Comment (jaspernbrouwer) [reply tests/Elewant/Webapp/Application/Controllers/ApiCommandAdoptElePHPantTest.php:31]: I'm all for being less strict in test-classes. Let's make a list of rules that can be suppressed for these!

File: tests/Elewant/Webapp/Application/Controllers/ApiCommandBase.php (hunk 16 above)
Comment (ramondelafuente) [reply tests/Elewant/Webapp/Application/Controllers/ApiCommandBase.php:164]: Oh yeah, this one was uncalled for...

When I wrote the comment I assumed it was another code-style "rule" that there must be a dot after every sentence; but it was probably just a manual addition without an actual improvement in mind?
